### ISSUE TITLE
show laylines to waypoint

### ISSUE THREAD

**Issue by petsicola** [SignalK/freeboard-sk#149]
show laylines to waypoint
![image](https://github.com/SignalK/freeboard-sk/assets/61534583/4ddd52a0-998b-4c4a-8ae1-5eb5fe5167ab)

**Comment by panaaj** [comment]
Available in v2.8.0

**Comment by petsicola** [comment]
Laylines should only be visible when the waypoint is upwind. If the heading to the waypoint is direct, they should disappear. do you think that?

**Comment by petsicola** [comment]
I think that I would show them until the wind exceeds 90 degrees

**Comment by panaaj** [comment]
Fixed in v2.8.2

**Comment by petsicola** [comment]
About the laylines. in dashT tactics plugin explain the problem of laylines. 
1º Laylines are displayed until the direct route is faster than the laylines. This is calculated with the polars. you can see it in the tactics plugin doc. https://dashboard-tactics-pi.readthedocs.io/en/latest/tactics/tactics.html
2º  I think that in our case the laylines should be shown when the route to the waypoint is upwind, not when the boat is upwind. Anyway, I'll give you an example later.

**Comment by panaaj** [comment]
Will fix 2. display when destination is upwind (not the vessel)

**Comment by petsicola** [comment]
I think there is an Issue, sorry... I think that laylines are in downwind

No lay lines
![image](https://github.com/SignalK/freeboard-sk/assets/61534583/16ae0eba-6477-4e34-bbcf-9d96775a1a7b)

Yes lay lines
![image](https://github.com/SignalK/freeboard-sk/assets/61534583/c7913064-8226-4cfe-b97c-17a53fee843d)

**Comment by petsicola** [comment]
With an image... when the green point is inside the two white lines you have to show the laylines. and as I said in the other message this lines are configured by the user in mxtommy kip.

![image](https://github.com/SignalK/freeboard-sk/assets/61534583/0981100e-c3c0-4e39-a1cd-77eb1727fe39)

**Comment by panaaj** [comment]
@petsicola The Close Haul value in KIP settings is not available to Freeboard-SK, so it cannot be used to determine when to display the laylines. 
This value would need to be added as a setting in Freeboard-SK and set to match the value entered into KIP.
Ideally this would be sourced from a Signal K path (e.g. polars) but one can't guarantee that the path will be populated with a value.

So, for clarity moving forward, I am going to propose that..... **Laylines will be displayed when ALL of the following conditions are met:**
1. A destination location is set
2. The difference in value between Bearing to Destination and True Wind Direction is less (TACK_ANGLE / 2) _(where the default TACK_ANGLE = 90)_

_Note: These rules will apply regardless of the vessel heading._

If these rules are applied will this meet your use case?

**Comment by petsicola** [comment]
Yes!
only one thing I don't know if it is a problem... what happens if true direction wind angle is 5º and bearing to destination is set at 359º the difference are 6º but 359-5=354º

if that is a problem... I don't know.... Henry in that discord gives a solution in node red and I think that we have to make something similar...
https://discord.com/channels/1170433917761892493/1170456013552496730/1238797469098180618

sorry I have a poor english and we are talking by github.

**Comment by tkurki** [comment]
Some additional thoughts - there are several different things at play here.

>Laylines represent the most efficient way of reaching your destination, given the presentwind conditions

If you have a **destination / mark set** and you know the boat's (present condition) **upwind target angles** you can show **Upwind Mark Laylines**. Same for **downwind**. These are *not dependent on where the boat is pointing*, only on the boat and mark's positions relative to the true wind direction. So when the mark is upwind of the boat the laylines are drawn downwind from the mark and vice versa if the mark is downwind.

**Boat laylines** are not dependent on the destination being set, as they are just a function of the present true wind angle and the boat's upwind and downwind target angles. **Boat laylines** are drawn upwind or downwind from the boat, depending on where the boat is going (courseOverGround I think).

If you have both Boat laylines and Mark Laylines available where they meet is the *Tack point*. Drawing the laylines for both tacks  just up to the tack point results in the 4 line shape.

https://chsmith.com.au/news/what-are-laylines-and-how-do-they-work-2013-04-24-8-31-07.html

To summarise:
- boat laylines are there (or can be drawn) always, all that is needed is boat's target angle and true wind direction
- mark laylines need destination & boat position, true wind direction and boat's target angle

Second point:
To make Freeboard and Kip play together is for both to use boat target angles, either as fixed values or calculated dynamically from polar data.

**Comment by godind** [comment]
I concur with this explanation. To remove the confusion around the term layline, in the tacking angle settings (be it fixed or from a path), I prefer the term Close-hauled Angle when referring to the boat's upwind target angle.

Using perf and polar will clear this. We are not there yet. but the data is!

**Comment by petsicola** [comment]
which is the path that I have to use to show lay lines???

**Comment by panaaj** [comment]
They will show when bearing is +/- `performance.beatAngle` (or 45 deg if there is no value) of true wind.

**Comment by petsicola** [comment]
Laylines doesn´t appear always. in some casses. doesnt appear.

**Comment by petsicola** [comment]
In this position laylines... I think that the have to appear. 
![image](https://github.com/SignalK/freeboard-sk/assets/61534583/64f703ce-670a-4de8-83ac-c65e179a1f5c)
